File: KOTLIN_PROTOTYPE/src/main/kotlin/LibraryManager.kt
import javafx.application.Platform
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.layout.BorderPane
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.stage.Modality
import javafx.stage.Stage
import java.io.File
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

// Models
data class Book
(
    val title:           String,
    val sentenceCount:   Int,
    val filePath:        String,
    val addedDate:       LocalDateTime,
    val originalPdfName: String
)

data class Checkpoint
(
    val bookTitle:     String,
    val sentenceIndex: Int,
    val savedDate:     LocalDateTime,
    val filePath:      String
)

class LibraryManager
{
    // Library collections
    val library:     ObservableList<Book>       = FXCollections.observableArrayList()
    val checkpoints: ObservableList<Checkpoint> = FXCollections.observableArrayList()
    
    private val libraryDir = File(System.getProperty("user.home"), ".pdfteller_library")

    init
    {
        if (!libraryDir.exists())
            libraryDir.mkdirs() // Create library directory if it doesn't exist...

        loadLibrary()
        loadCheckpoints()
    }

    fun saveBookToLibrary(title: String, sentences: List<String>, originalPdfName: String): Boolean
    {
        // Check if book already exists (same title and sentence count)
        val existingBook = library.find {
            it.title == title && it.sentenceCount == sentences.size
        }

        if (existingBook != null)
            return false // Book already exists, don't save again

        // Create unique filename
        val timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"))
        val fileName  = "${title}_${timestamp}.txt"
        val filePath  = File(libraryDir, fileName).absolutePath

        // Save sentences to file
        return try
        {
            File(filePath).writeText(sentences.joinToString("\n"))

            // Add to library
            val book = Book(
                title           = title,
                sentenceCount   = sentences.size,
                filePath        = filePath,
                addedDate       = LocalDateTime.now(),
                originalPdfName = originalPdfName
            )

            Platform.runLater {
                library.add(book)
                saveLibraryMetadata()
            }
            true
        }
        catch (e: Exception)
        {
            println(e)
            false
        }
    }

    fun saveCheckpoint(bookTitle: String, sentenceIndex: Int, filePath: String): Boolean
    {
        return try
        {
            val checkpoint = Checkpoint(
                bookTitle     = bookTitle,
                sentenceIndex = sentenceIndex,
                savedDate     = LocalDateTime.now(),
                filePath      = filePath
            )

            // Remove old checkpoints for the same book
            checkpoints.removeIf { it.bookTitle == bookTitle }
            checkpoints.add(checkpoint)
            saveCheckpointsMetadata()
            true
        }
        catch (e: Exception)
        {
            println(e)
            false
        }
    }

    private fun loadLibrary()
    {
        val metadataFile = File(libraryDir, "library.json")
        if (!metadataFile.exists()) return

        try
        {
            val lines = metadataFile.readLines()
            lines.forEach { line ->
                if (line.contains("title:"))
                {
                    val parts = line.split("|")
                    if (parts.size >= 5)
                    {
                        val book = Book(
                            title = parts[0].substringAfter("title:"),
                            sentenceCount = parts[1].substringAfter("count:").toInt(),
                            filePath = parts[2].substringAfter("path:"),
                            addedDate = LocalDateTime.parse(parts[3].substringAfter("date:")),
                            originalPdfName = parts[4].substringAfter("pdf:")
                        )
                        if (File(book.filePath).exists())
                            library.add(book)
                    }
                }
            }
        }
        catch (e: Exception) { println(e) }
    }

    private fun saveLibraryMetadata()
    {
        val metadataFile = File(libraryDir, "library.json")
        try
        {
            val content = library.joinToString("\n") { book ->
                "title:${book.title}|count:${book.sentenceCount}|path:${book.filePath}|date:${book.addedDate}|pdf:${book.originalPdfName}"
            }
            metadataFile.writeText(content)
        }
        catch (e: Exception) { println(e) }
    }

    private fun loadCheckpoints()
    {
        val checkpointsFile = File(libraryDir, "checkpoints.txt")
        if (!checkpointsFile.exists()) return

        try
        {
            val lines = checkpointsFile.readLines()
            lines.forEach { line ->
                if (line.contains("title:"))
                {
                    val parts = line.split("|")
                    if (parts.size >= 4)
                    {
                        val checkpoint = Checkpoint(
                            bookTitle     = parts[0].substringAfter("title:"),
                            sentenceIndex = parts[1].substringAfter("index:").toInt(),
                            savedDate     = LocalDateTime.parse(parts[2].substringAfter("date:")),
                            filePath      = parts[3].substringAfter("path:")
                        )
                        if (File(checkpoint.filePath).exists())
                            checkpoints.add(checkpoint)
                    }
                }
            }
        }
        catch (e: Exception) { println(e) }
    }

    private fun saveCheckpointsMetadata()
    {
        val checkpointsFile = File(libraryDir, "checkpoints.txt")
        try
        {
            val content = checkpoints.joinToString("\n") { checkpoint ->
                "title:${checkpoint.bookTitle}|index:${checkpoint.sentenceIndex}|date:${checkpoint.savedDate}|path:${checkpoint.filePath}"
            }
            checkpointsFile.writeText(content)
        }
        catch (e: Exception) { println(e) }
    }

    fun loadBookFromLibrary(book: Book): List<String>?
    {
        return try
        {
            File(book.filePath).readLines()
        }
        catch (e: Exception)
        {
            println(e)
            null
        }
    }

    fun loadCheckpoint(checkpoint: Checkpoint): List<String>?
    {
        return try
        {
            File(checkpoint.filePath).readLines()
        }
        catch (e: Exception)
        {
            println(e)
            null
        }
    }

    fun deleteBook(book: Book): Boolean
    {
        return try
        {
            File(book.filePath).delete()
            library.remove(book)

            // Also remove associated checkpoints
            checkpoints.removeIf { it.bookTitle == book.title }

            saveLibraryMetadata()
            saveCheckpointsMetadata()
            
            true
        }
        catch (e: Exception)
        {
            println(e)
            false
        }
    }

    fun deleteCheckpoint(checkpoint: Checkpoint): Boolean
    {
        return try
        {
            checkpoints.remove(checkpoint)
            saveCheckpointsMetadata()
            true
        }
        catch (e: Exception)
        {
            println(e)
            false
        }
    }

    fun showLibrary(primaryStage: Stage, onBookLoaded: (Book, List<String>) -> Unit, onCheckpointLoaded: (Checkpoint, List<String>) -> Unit)
    {
        val dialog = Stage().apply {
            title = "Library"
            initModality(Modality.APPLICATION_MODAL)
            initOwner(primaryStage)
        }

        val tabPane = TabPane()

        // Books Tab
        val booksTab = Tab("Books").apply {
            isClosable = false
            content    = createBooksTab(dialog, onBookLoaded)
        }

        // Checkpoints Tab
        val checkpointsTab = Tab("Checkpoints").apply {
            isClosable = false
            content    = createCheckpointsTab(dialog, onCheckpointLoaded)
        }

        tabPane.tabs.addAll(booksTab, checkpointsTab)

        val root = BorderPane().apply {
            center  = tabPane
            padding = Insets(10.0)
        }

        dialog.scene = Scene(root, 500.0, 550.0)
        dialog.show()
    }

    private fun createBooksTab(dialog: Stage, onBookLoaded: (Book, List<String>) -> Unit): VBox
    {
        val listView = ListView(library).apply {
            setCellFactory {
                object : ListCell<Book>()
                {
                    override fun updateItem(item: Book?, empty: Boolean)
                    {
                        super.updateItem(item, empty)
                        if (empty || item == null)
                        {
                            text    = null
                            graphic = null
                        }
                        else
                        {
                            val vbox = VBox(5.0).apply {
                                children.addAll(
                                    Label(item.title).apply {
                                        style = "-fx-font-weight: bold; -fx-font-size: 14px;"
                                    },
                                    Label("${item.sentenceCount} sentences"),
                                    Label("Added: ${item.addedDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"))}")
                                )
                                padding = Insets(5.0)
                            }
                            graphic = vbox
                        }
                    }
                }
            }
            prefHeight = 400.0
        }

        val loadButton = Button("Load Selected").apply {
            prefWidth = 120.0
            setOnAction {
                val selected = listView.selectionModel.selectedItem
                if (selected != null)
                {
                    val sentences = loadBookFromLibrary(selected)
                    if (sentences != null)
                    {
                        onBookLoaded(selected, sentences)
                        dialog.close()
                    }
                }
            }
        }

        val deleteButton = Button("Delete").apply {
            prefWidth = 120.0
            setOnAction {
                val selected = listView.selectionModel.selectedItem
                if (selected != null)
                {
                    val alert = Alert(Alert.AlertType.CONFIRMATION).apply {
                        title       = "Delete Book"
                        headerText  = "Delete \"${selected.title}\"?"
                        contentText = "This action cannot be undone."
                    }

                    val result = alert.showAndWait()
                    if (result.isPresent && result.get() == ButtonType.OK)
                        deleteBook(selected)
                }
            }
        }

        val closeButton = Button("Close").apply {
            prefWidth = 120.0
            setOnAction { dialog.close() }
        }

        val buttonBar = HBox(10.0, loadButton, deleteButton, closeButton).apply {
            alignment = Pos.CENTER
            padding   = Insets(10.0)
        }

        return VBox().apply {
            children.addAll(listView, buttonBar)
            VBox.setVgrow(listView, javafx.scene.layout.Priority.ALWAYS)
        }
    }

    private fun createCheckpointsTab(dialog: Stage, onCheckpointLoaded: (Checkpoint, List<String>) -> Unit): VBox
    {
        val listView = ListView(checkpoints).apply {
            setCellFactory {
                object : ListCell<Checkpoint>()
                {
                    override fun updateItem(item: Checkpoint?, empty: Boolean)
                    {
                        super.updateItem(item, empty)
                        if (empty || item == null)
                        {
                            text    = null
                            graphic = null
                        }
                        else
                        {
                            val vbox = VBox(5.0).apply {
                                children.addAll(
                                    Label(item.bookTitle).apply {
                                        style = "-fx-font-weight: bold; -fx-font-size: 14px;"
                                    },
                                    Label("At sentence: ${item.sentenceIndex + 1}"),
                                    Label("Saved: ${item.savedDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"))}")
                                )
                                padding = Insets(5.0)
                            }
                            graphic = vbox
                        }
                    }
                }
            }
            prefHeight = 400.0
        }

        val loadButton = Button("Load Checkpoint").apply {
            prefWidth = 130.0
            setOnAction {
                val selected = listView.selectionModel.selectedItem
                if (selected != null)
                {
                    val sentences = loadCheckpoint(selected)
                    if (sentences != null)
                    {
                        onCheckpointLoaded(selected, sentences)
                        dialog.close()
                    }
                }
            }
        }

        val deleteButton = Button("Delete").apply {
            prefWidth = 120.0
            setOnAction {
                val selected = listView.selectionModel.selectedItem
                if (selected != null)
                    deleteCheckpoint(selected)
            }
        }

        val closeButton = Button("Close").apply {
            prefWidth = 120.0
            setOnAction { dialog.close() }
        }

        val buttonBar = HBox(10.0, loadButton, deleteButton, closeButton).apply {
            alignment = Pos.CENTER
            padding   = Insets(10.0)
        }

        return VBox().apply {
            children.addAll(listView, buttonBar)
            VBox.setVgrow(listView, javafx.scene.layout.Priority.ALWAYS)
        }
    }
}

File: PYTHON_PROTOTYPE/library_manager.py
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Optional
import json
import os


@dataclass
class Book:
    title: str
    sentence_count: int
    file_path: str
    added_date: datetime
    original_pdf_name: str


@dataclass
class Checkpoint:
    book_title: str
    sentence_index: int
    saved_date: datetime
    file_path: str


class LibraryManager:
    def __init__(self):
        self.library: list[Book] = []
        self.checkpoints: list[Checkpoint] = []

        self.library_dir = Path.home() / ".pdfteller_library"

        if not self.library_dir.exists():
            self.library_dir.mkdir(parents=True)

        self.load_library()
        self.load_checkpoints()

    def _parse_datetime(self, date_str: str) -> datetime:
        """Parse datetime string, handling high-precision microseconds"""
        # Handle microseconds with more than 6 digits
        if '.' in date_str:
            base, microseconds = date_str.rsplit('.', 1)
            # Truncate microseconds to 6 digits
            microseconds = microseconds[:6]
            date_str = f"{base}.{microseconds}"
        return datetime.fromisoformat(date_str)

    def save_book_to_library(self, title: str, sentences: list[str], original_pdf_name: str) -> bool:
        # Check if book already exists (same title and sentence count)
        existing_book = next(
            (book for book in self.library
             if book.title == title and book.sentence_count == len(sentences)),
            None
        )

        if existing_book is not None:
            return False  # Book already exists

        # Create unique filename
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        file_name = f"{title}_{timestamp}.txt"
        file_path = str(self.library_dir / file_name)

        # Save sentences to file
        try:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write('\n'.join(sentences))

            # Add to library
            book = Book(
                title=title,
                sentence_count=len(sentences),
                file_path=file_path,
                added_date=datetime.now(),
                original_pdf_name=original_pdf_name
            )

            self.library.append(book)
            self.save_library_metadata()
            return True
        except Exception as e:
            print(f"Error saving book: {e}")
            return False

    def save_checkpoint(self, book_title: str, sentence_index: int, file_path: str) -> bool:
        try:
            checkpoint = Checkpoint(
                book_title=book_title,
                sentence_index=sentence_index,
                saved_date=datetime.now(),
                file_path=file_path
            )

            # Remove old checkpoints for the same book
            self.checkpoints = [cp for cp in self.checkpoints if cp.book_title != book_title]
            self.checkpoints.append(checkpoint)
            self.save_checkpoints_metadata()
            return True
        except Exception as e:
            print(f"Error saving checkpoint: {e}")
            return False

    def load_library(self):
        metadata_file = self.library_dir / "library.json"
        if not metadata_file.exists():
            return

        try:
            with open(metadata_file, 'r', encoding='utf-8') as f:
                lines = f.readlines()

            for line in lines:
                if "title:" in line:
                    parts = line.strip().split("|")
                    if len(parts) >= 5:
                        book = Book(
                            title=parts[0].split("title:")[1],
                            sentence_count=int(parts[1].split("count:")[1]),
                            file_path=parts[2].split("path:")[1],
                            added_date=self._parse_datetime(parts[3].split("date:")[1]),
                            original_pdf_name=parts[4].split("pdf:")[1]
                        )
                        if Path(book.file_path).exists():
                            self.library.append(book)
        except Exception as e:
            print(f"Error loading library: {e}")

    def save_library_metadata(self):
        metadata_file = self.library_dir / "library.json"
        try:
            content = '\n'.join([
                f"title:{book.title}|count:{book.sentence_count}|path:{book.file_path}|date:{book.added_date.isoformat()}|pdf:{book.original_pdf_name}"
                for book in self.library
            ])
            with open(metadata_file, 'w', encoding='utf-8') as f:
                f.write(content)
        except Exception as e:
            print(f"Error saving library metadata: {e}")

    def load_checkpoints(self):
        checkpoints_file = self.library_dir / "checkpoints.txt"
        if not checkpoints_file.exists():
            return

        try:
            with open(checkpoints_file, 'r', encoding='utf-8') as f:
                lines = f.readlines()

            for line in lines:
                if "title:" in line:
                    parts = line.strip().split("|")
                    if len(parts) >= 4:
                        checkpoint = Checkpoint(
                            book_title=parts[0].split("title:")[1],
                            sentence_index=int(parts[1].split("index:")[1]),
                            saved_date=self._parse_datetime(parts[2].split("date:")[1]),
                            file_path=parts[3].split("path:")[1]
                        )
                        if Path(checkpoint.file_path).exists():
                            self.checkpoints.append(checkpoint)
        except Exception as e:
            print(f"Error loading checkpoints: {e}")

    def save_checkpoints_metadata(self):
        checkpoints_file = self.library_dir / "checkpoints.txt"
        try:
            content = '\n'.join([
                f"title:{cp.book_title}|index:{cp.sentence_index}|date:{cp.saved_date.isoformat()}|path:{cp.file_path}"
                for cp in self.checkpoints
            ])
            with open(checkpoints_file, 'w', encoding='utf-8') as f:
                f.write(content)
        except Exception as e:
            print(f"Error saving checkpoints metadata: {e}")

    def load_book_from_library(self, book: Book) -> Optional[list[str]]:
        try:
            with open(book.file_path, 'r', encoding='utf-8') as f:
                return f.read().splitlines()
        except Exception as e:
            print(f"Error loading book: {e}")
            return None

    def load_checkpoint(self, checkpoint: Checkpoint) -> Optional[list[str]]:
        try:
            with open(checkpoint.file_path, 'r', encoding='utf-8') as f:
                return f.read().splitlines()
        except Exception as e:
            print(f"Error loading checkpoint: {e}")
            return None

    def delete_book(self, book: Book) -> bool:
        try:
            Path(book.file_path).unlink()
            self.library.remove(book)

            # Also remove associated checkpoints
            self.checkpoints = [cp for cp in self.checkpoints if cp.book_title != book.title]

            self.save_library_metadata()
            self.save_checkpoints_metadata()
            return True
        except Exception as e:
            print(f"Error deleting book: {e}")
            return False

    def delete_checkpoint(self, checkpoint: Checkpoint) -> bool:
        try:
            self.checkpoints.remove(checkpoint)
            self.save_checkpoints_metadata()
            return True
        except Exception as e:
            print(f"Error deleting checkpoint: {e}")
            return False
